src/math/mod.rs

```diff
@@ -29,6 +29,124 @@ pub fn canon_egcd(a: i64, b: i64, c: i64) -> Option<(i64, i64, i64)> {
     }
 }
 
+fn pos_mod(n: i64, m: i64) -> i64 {
+    if n < 0 {
+        n + m
+    } else {
+        n
+    }
+}
+
+fn mod_mul(a: i64, b: i64, m: i64) -> i64 {
+    pos_mod((a as i128 * b as i128 % m as i128) as i64, m)
+}
+
+/// Assuming m >= 2 and exp >= 0, finds base ^ exp % m in logarithmic time
+pub fn mod_exp(mut base: i64, mut exp: i64, m: i64) -> i64 {
+    assert!(m >= 2);
+    assert!(exp >= 0);
+    let mut ans = 1 % m;
+    base = base % m;
+    while exp > 0 {
+        if exp % 2 == 1 {
+            ans = mod_mul(ans, base, m);
+        }
+        base = mod_mul(base, base, m);
+        exp /= 2;
+    }
+    pos_mod(ans, m)
+}
+
+/// Assuming m >= 2, finds multiplicative inverse of n under modulus m
+pub fn mod_inv(n: i64, m: i64) -> i64 {
+    mod_exp(n, m - 2, m)
+}
+
+fn miller_test(n: i64, d: i64, r: i64, a: i64) -> bool {
+    let mut x = mod_exp(a, d, n);
+    if x == 1 || x == n - 1 {
```

Comment:
Might be clearer if this `x == n-1` check is removed in favor of checking it at the beginning of the `for` loop.

Reply:
I will leave that up to you, as far as I know most implementations I've seen do this check the same way I did it. I also feel like it's easier to understand.

Reply:
I kinda like the shortened way, not a big deal though. Note that your current implementation checks against `n-1` r+1 times, whereas the Wikipedia version checks r times.

Btw, it might be clearer to name this function something like `is_strong_probable_prime` to say something about its return value.